core/dbt/events/types.py

```diff
@@ -636,15 +638,210 @@ class InvalidDisabledSourceInTestNode(WarnLevel, CliEventABC):
+        border = '-' * len(msg)
+        return f"  See test failures:\n  {border}\n  {msg}\n  {border}"
+
+
+@dataclass
+class FirstRunResultError(ErrorLevel, CliEventABC):
+    msg: str
+
+    def cli_msg(self) -> str:
+        return ui.yellow(self.msg)
+
+
+@dataclass
+class AfterFirstRunResultError(ErrorLevel, CliEventABC):
+    msg: str
+
+    def cli_msg(self) -> str:
+        return self.msg
+
+
+@dataclass
+class EndOfRunSummary(InfoLevel, CliEventABC):
+    num_errors: int
+    num_warnings: int
+    keyboard_interrupt: bool = False
+
+    def cli_msg(self) -> str:
+        error_plural = utils.pluralize(self.num_errors, 'error')
+        warn_plural = utils.pluralize(self.num_warnings, 'warning')
+        if self.keyboard_interrupt:
+            message = ui.yellow('Exited because of keyboard interrupt.')
+        elif self.num_errors > 0:
+            message = ui.red("Completed with {} and {}:".format(
+                error_plural, warn_plural))
+        elif self.num_warnings > 0:
+            message = ui.yellow('Completed with {}:'.format(warn_plural))
+        else:
+            message = ui.green('Completed successfully')
+        return message
+
+
+def format_fancy_output_line(
+        msg: str, status: str, index: Optional[int],
+        total: Optional[int], execution_time: Optional[float] = None,
+        truncate: bool = False
+) -> None:
+    if index is None or total is None:
+        progress = ''
+    else:
+        progress = '{} of {} '.format(index, total)
+    prefix = "{timestamp} | {progress}{message}".format(
+        timestamp=get_timestamp(),
+        progress=progress,
+        message=msg)
+
+    truncate_width = ui.printer_width() - 3
+    justified = prefix.ljust(ui.printer_width(), ".")
+    if truncate and len(justified) > truncate_width:
+        justified = justified[:truncate_width] + '...'
+
+    if execution_time is None:
+        status_time = ""
+    else:
+        status_time = " in {execution_time:0.2f}s".format(
+            execution_time=execution_time)
+
+    output = "{justified} [{status}{status_time}]".format(
+        justified=justified, status=status, status_time=status_time)
+
+    return output
+
+
+@dataclass
+class PrintStartLine(InfoLevel, CliEventABC):
+    description: str
+    index: int
+    total: int
+
+    def cli_msg(self) -> str:
+        msg = "START {self.description}"
+        return format_fancy_output_line(msg=msg, status='RUN', index=self.index, total=self.total)
+
+
+@dataclass
+class PrintHookStartLine(InfoLevel, CliEventABC):
+    statement: str
+    index: int
+    total: int
+    truncate: bool
+
+    def cli_msg(self) -> str:
+        msg = "START hook: {self.statement}"
+        return format_fancy_output_line(msg=msg,
+                                        status='RUN',
+                                        index=self.index,
+                                        total=self.total,
+                                        truncate=self.truncate)
```

Comment:
An alternative to this approach of one type per message is a general log statement for each level: Error, Warn and Info that is then shared across messages in `core/dbt/task/printer.py`.  That goes against what we've been doing elsewhere though.

Reply:
Yeah that's an alternative I considered. It doesn't really solve the problem of having development time notice when our public interface changes. Leveraging mypy via these types gives us a little of that even though it might look a bit messy at times.